elasticapm/utils/encoding.py

```diff
@@ -178,15 +178,41 @@ def to_string(value):
         return to_unicode(value).encode("utf-8")
 
 
-def shorten(var, list_length=50, string_length=200):
+def shorten(var, list_length=50, string_length=200, dict_length=50):
+    """
+    Shorten a given variable based on configurable maximum lengths, leaving
+    breadcrumbs in the object to show that it was shortened.
+
+    For strings, truncate the string to the max length, and append "..." so
+    the user knows data was lost.
+
+    For lists, truncate the list to the max length, and append two new strings
+    to the list: "..." and "(<x> more elements)" where <x> is the number of
+    elements removed.
+
+    For dicts, truncate the dict to the max length (based on number of key/value
+    pairs) and add a new (key, value) pair to the dict:
+    ("...", "(<x> more elements)") where <x> is the number of key/value pairs
+    removed.
+
+    :param var: Variable to be shortened
+    :param list_length: Max length (in items) of lists
+    :param string_length: Max length (in characters) of strings
+    :param dict_length: Max length (in key/value pairs) of dicts
+    :return: Shortened variable
+    """
     var = transform(var)
     if isinstance(var, compat.string_types) and len(var) > string_length:
         var = var[: string_length - 3] + "..."
     elif isinstance(var, (list, tuple, set, frozenset)) and len(var) > list_length:
         # TODO: we should write a real API for storing some metadata with vars when
         # we get around to doing ref storage
-        # TODO: when we finish the above, we should also implement this for dicts
         var = list(var)[:list_length] + ["...", "(%d more elements)" % (len(var) - list_length,)]
+    elif isinstance(var, dict) and len(var) > dict_length:
+        trimmed_tuples = [(k, v) for (k, v) in compat.iteritems(var)][:dict_length]
+        if "..." not in var:
+            trimmed_tuples += [("...", "(%d more elements)" % (len(var) - dict_length))]
```

Comment:
Since this is adding an extra element to the dict. If ones calls `shorten` twice, this element gets removed.  So, `shorten(long_dict)` will include the `...` key but `shorten(shorten(long_dict))` will not.
In my rudimentary tests, I've seen `varmap` being called twice for dicts. So the `...` key may not be present in the final message sent to the server.
A possible solution is to pop to replace the last element of the array with the dots instead of adding an extra element.
```suggestion
            trimmed_tuples[-1] = ("...", "(%d more elements)" % (len(var) - dict_length + 1))
```

Reply:
It's being called twice, but I think once with `_sanitize` and once with `shorten` -- that's why it was important to apply the patch here, instead of in `varmap` like in your monkey-patch on your issue. `varmap` is not a single-use mapping function, unfortunately.